### DISCUSSION TITLE
Reduce verbosity and pain for working with JSON in C#

### DISCUSSION THREAD

**Discussion by Nefcanto** [dotnet/csharplang#8535] [Language Ideas] [answered]
Consider this JSON:

```
{
  "return": {
    "status": 200,
    "message": "Approved"
  },
  "entries": [
    {
      "messageId": 1466447495,
      "message": "Your verification code: 23298",
      "status": 5,
      "statusText": "Sent",
      "sender": "1000551451",
      "receptor": "918181818181818181",
      "date": 1701615164
    }
  ]
}
```

Let's check if it has entries property, and the entries property is array and it has at least one element, then get the message property if it exists.

This is the JS code:

```
if (json.entries && json.entries.map && json.entries.length > 0) {
    const message = json.entries[0]?.message
}
```

This is the Python code:

```
_json = json.loads(json_string)
if 'entries' in _json and isinstance(_json['entries'], list) and len(_json['entries']) > 0:
    first_entry = _json['entries'][0]
    if 'message' in first_entry:
        message = first_entry['message']
```

More verbose, but still bearable.

This is the C# code:

```
var json = JsonSerializer.Deserialize<dynamic>(jsonString);
if (json is JsonElement jsonElement && jsonElement.TryGetProperty("entries", out JsonElement entries) &&  entries.ValueKind == JsonValueKind.Array && entries.GetArrayLength() > 0)
{
    var firstEntry = entries[0];
    if (firstEntry.TryGetProperty("message", out JsonElement message))
    {
        var message = message.GetString();
    }
}
```

See the amount of verbosity? More lines of code and less readability.

Please make working with JSON less verbose. Please enhance the DX in working with JSON in C#. It's a pain.
Make it a first-class citizen. Add some keywords.

This is a pseudo-code that could be valuable:

```
var json = ParseJson(jsongString);
if (json.Has("entries") && json["entries"].IsArray() && json["entries"].Length > 0)
{
    var message = json["entries"][0]?.["message"];
}
```

**Comment by Nefcanto** [comment]
I encourage others to add samples from more languages to see how bad C# is in this regard.

I asked ChatGPT and these are some more code examples:

Ruby:

```
json_object = JSON.parse(json_string)
if json_object.key?('entries') && json_object['entries'].any?
  first_entry = json_object['entries'][0]
  if first_entry.key?('message')
    message = first_entry['message']
    puts message
  end
end
```

PHP:

```
$jsonObject = json_decode($jsonString, true);
if (isset($jsonObject['entries']) && count($jsonObject['entries']) > 0) {
    $firstEntry = $jsonObject['entries'][0];
    if (isset($firstEntry['message'])) {
        $message = $firstEntry['message'];
        echo $message;
    }
}
```

Dart:

```
var jsonObject = jsonDecode(jsonString);
if (jsonObject.containsKey('entries') && jsonObject['entries'].isNotEmpty) {
var firstEntry = jsonObject['entries'][0];
if (firstEntry.containsKey('message')) {
    var message = firstEntry['message'];
    print(message);
}
}
```

Kotlin:

```
val jsonObject = JSONObject(jsonString)
if (jsonObject.has("entries")) {
    val entries = jsonObject.getJSONArray("entries")
    if (entries.length() > 0 && entries.getJSONObject(0).has("message")) {
        val message = entries.getJSONObject(0).getString("message")
        println(message)
    }
}
```

I don't know if they are valid codes or not. I'm not an expert in those languages. But if they are valid, then C# is truly awful in handling JSON.

**Reply by chrisoverzero** [reply]
Have you considered deserializing the response to a static type?

```c#
sealed record class Response(Response.Metadata Return, List<Response.Entry> Entries)
{
    public sealed record class Metadata(int Status, string Message);

    public sealed record class Entry(int MessageId, string Message, int Status, string StatusText, string Sender, string Receptor, int Date);
}

[JsonSerializable(typeof(Response))]
[JsonSourceGenerationOptions(Web)]
sealed partial class ExampleContext
    : JsonSerializerContext;
```

Then you can:

```c#
var resp = JsonSerializer.Deserialize(detail, ExampleContext.Default.Response);
if (resp is not { Entries: [{ Message: { } msg }, .. _] })
{
    Console.WriteLine("nope");
    return;
}

Console.WriteLine(msg);
```

Even if you want to keep it dynamic for some reason, there's no need to deserialize to the _type_ `dynamic`:

```c#
var reader = new Utf8JsonReader(detail);
var resp = JsonElement.ParseValue(ref reader);

// or

var resp = JsonNode.Parse(detail, new() { PropertyNameCaseInsensitive = true });
```

**Reply by huoyaoyuan** [reply]
I can't imagine how you classify the `if` chains as good. These languages are all terrible at testing nested objects.

**Reply by Nefcanto** [reply]
@chrisoverzero, are you inviting me to use more verbosity and boilerplate code? Really? Do you expect us to create a static type for any JSON we receive? That's the most boilerplated, verbose answer I've ever seen.

**Reply by Nefcanto** [reply]
@huoyaoyuan I'm using Python and I'm using C# on a daily basis. I must disagree with you. I always analyze AST (Abstract Syntax Trees) with Python and it's awesome even when it comes to testing nested objects. What is wrong with having multiple nested ifs?

**Reply by CyrusNajmabadi** [reply]
> @chrisoverzero, are you inviting me to use more verbosity and boilerplate code? Really? Do you expect us to create a static type for any JSON we receive? That's the most boilerplated, verbose answer I've ever seen.

I don't see the problem.  This is what we do in typescript as well. It allows the shape of the data to be described, ensuring type safety and catching of mistakes. 

For example, it's trivial to accidentally write "massage" instead of "message" when having to do everything in a weakly typed fashion. 

It's trivial to just write a simple your describing the data, and then use all the static features of c# to with with it. 

Play to the strengths of the system. This is what all those other languages would recommend as well.

**Reply by CyrusNajmabadi** [reply]
> What is wrong with having multiple nested ifs? 

> That's the most boilerplated, verbose answer I've ever seen.

:-)

C# is not Python. It does not aim to be. Your dynamic testing approach makes sense there where everything is dynamic. C# and .net excel in performance and efficiency in static approaches to data processing. Patterns nicely ensure you match a shape while simultaneously extracting the data you need. 

If you avoid the strengths of the system and blindly insist on trying to some problems in the weak way that you do so elsewhere, your going to have a bad time. 

Finally, we are interested in making these spaces better. But we do so by investing in ways that some the issue holistically. We don't just ape what another language has done as there are often much better ways available to us based on how our entire runtime and type system are designed.

**Comment by Unknown6656** [comment]
This sounds like you're requesting a library change, not a language / syntax change. I.e., you want the `System.Text.Json.*`-APIs to be less verbose, but you're not requesting some explicit C# syntax changes, am I correct?

Should that be the case, I'd advise you to forward your question/issue to the following repo:
https://github.com/dotnet/runtime

**Reply by HaloFour** [reply]
A couple of extension methods could probably wrap all of that up into a single call, but either way, this is a library concern, not a language one.

**Reply by HaloFour** [reply]
Shouldn't be too hard to accomplish the following with extension methods:

```cs
var jsonObject = JsonObject.Parse(...);
if (jsonObject.HasArray("entries", out var entries) && entries.Length > 0) {
    var message = entries[0].GetString("message");
}
```

**Comment by KennethHoff** [comment]
I think active patterns would solve this quite nicely. I fail to find the proposal though..

Being able to do something like this:

```cs
var json = ParseJson(jsongString);
if (json is JsonArray { entries: [{} firstEntry] } && firstEntry.TryGetProperty("message", out JsonElement message))
{
   var message = message.GetString();
}
```

Where `JsonArray` is defined something like this [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWAChiA3CZAAgCsBBZZCAT0oF5KYBTAd0oCkAzgHsY9RkwAUAbTxwcAXQCUAbjIVKnMAAcELdl16CRY5tKWqSlqADNKEug2aUoA/sNGOWAb0oBRADaanDAIAiAcQgiUUvKUEDAAJpTkVNQBQSECAL6KlF7ElJRYeACcEgAkAER84V5pgWDBoQB0ADLBAOYIABZZlSrEWcSc/gKceQWFU8VllQByQm79FkOWxDZ2mjosLm7Gnnl+DU1hEVExcYnJ6pzpjZk5E0SFMxWVvnm3x5ltnT19AyIqxGYyeU2mpQk80WvgEy0GxGIWAAzEUcHsPOIJLAEBcRhlQrl8s8iqjkJwIAkRP4diELncTmwNN9QisgA)

**Reply by jnm2** [reply]
Proposal for active patterns, aka user-defined positional patterns: https://github.com/dotnet/csharplang/issues/4131

**Comment by AoshiW** [comment]
I think the `JsonNode` is the solution to your problem.
```cs
using System.Text.Json;
using System.Text.Json.Nodes;

var json = JsonSerializer.Deserialize<JsonNode>(jsonString);
if (json["entries"] is JsonArray array && array.Count > 0)
{
    var message = array[0]["message"].GetValue<string>();
}
```

**Reply by jnm2** [reply]
Why not `var json = JsonNode.Parse(jsonString);`?

**Reply by AoshiW** [reply]
I didn't use it because I forgot that method existed.

**Comment by koszeggy** [comment]
I doubt that the existing API would introduce such breaking changes but you can use 3rd party [solutions](https://github.com/koszeggy/KGySoft.Json) if reducing verbosity is really that important (disclaimer: I am the author, so I hope this is not qualified as an unwanted advertisement).

With the library above the solution is this compact (see also [online](https://dotnetfiddle.net/9ZSMvt)):

```cs
var json = JsonValue.Parse(jsonString);
string message = json["entries"][0]["message"].AsString ?? throw new ArgumentException("Invalid content");
```

You can do this because getting lost in the hierarchy just returns a default `JsonValue`, which is the equivalent of JS `undefined`:

```cs
Console.WriteLine(json["entrdries"][-42]["msg"]); // undefined
```

And [here](https://github.com/koszeggy/KGySoft.Json/blob/master/KGySoft.Json.PerformanceTest/PerformanceTests/ReadDomTest.cs#L77-L81) you can compare the verbosity of the different libraries (well, this is when there is no validation; otherwise, other libraries are just even more verbose).

**Comment by Trarizon** [comment]
How about a struct wrapper of `JsonElement` like
``` c#
public readonly struct JsonElementWrapper(JsonElement element)
{
    public JsonElementWrapper? this[int index] => element.ValueKind is JsonValueKind.Array ? new(element[index]) : null;
    public JsonElementWrapper? this[ReadOnlySpan<char> propertyName] => element.ValueKind is JsonValueKind.Object ? new(element.GetProperty(propertyName)) : null;
    public int? ArrayLength => element.ValueKind is JsonValueKind.Array ? element.GetArrayLength() : null;
    public string? AsString() => element.ValueKind is JsonValueKind.String ? element.GetString() : null;
}
```
and you can write
``` c#
using var doc = JsonDocument.Parse(jsonString);
var root = new JsonElementWrapper(doc.RootElement);

if (root["entries"] is { ArrayLength: > 0 } entries) {
    var message = entries[0]?["message"]?.AsString();
}
```

**Comment by zms9110750** [comment]
I think you think it's long because you used the wrong API.
```csharp
using Newtonsoft.Json.Linq; 

JObject json = JObject.Parse(jsonString); 
 
if (json["entries"] is JArray { Count: > 0 } jar)
{
	Console.WriteLine(jar[0]["message"]);
} 
```

**Reply by huoyaoyuan** [reply, marked as answer]
This can be further simplified with list pattern:

```csharp
if (json["entries"] is JArray and [first, ..])
{
	Console.WriteLine(first["message"]);
} 
```

**Reply by zms9110750** [reply]
@huoyaoyuan  can't.
`Unhandled exception. System.ArgumentException: Accessed JArray values with invalid key value: 0. Int32 array index expected.`
I don't know this is API issue or lang issue.

**Reply by CyrusNajmabadi** [reply]
> @huoyaoyuan  can't.
> `Unhandled exception. System.ArgumentException: Accessed JArray values with invalid key value: 0. Int32 array index expected.`
> I don't know this is API issue or lang issue.

It's an API issue. The lang provides the constructs, it's up the API to integrate with them.

**Reply by elgonzo** [reply]
@CyrusNajmabadi 
> The lang provides the constructs, it's up the API to integrate with them.

Then how is an API supposed to integrate with this? Considering that the exception is thrown on an `JArray` instance which implements `IList<JToken>` and has an `public JToken this[int index]` indexer, i don't see what else would be required for the list pattern to work...

**Reply by zms9110750** [reply]
@elgonzo 
> JToken this[int index]

no, it's `JToken this[object key]`.
and it have `if ( key not is int) throw`
so,list pattern  give it `Index`

**Reply by elgonzo** [reply]
@zms9110750 look at this: https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_Linq_JArray_Item.htm
(A further hint is `JArray` being a `IList<JToken>`, so it therefore _must_ have an int-based indexer...)

If the compiler is unable to pick the existing indexer befitting the list pattern, then this would be a language/compiler issue, not an API issue.

As it seems, the compiler is just happily boxing an `Index` instance and passing it to the object-based indexer, assuming everything will be alright. Which in my opinion is a rather bold assumption to make, assuming that an `object`-based indexer (which is not part of the IList contract) of any arbitrary type (like JArray) is somehow more suitable than the `int`-based indexer wich is actually part of the IList contract.

**Reply by zms9110750** [reply]
@elgonzo ok.
The document states that 
>when the compiler automatically synthesizes' this [index] ', you need to ensure that you do not have a true' this [index] '`

now, Let's add one sentence: there is no `this [object]`

>If the compiler is unable to pick the existing indexer befitting the list pattern, then this would be a language/compiler issue, not an API issue...

I agree

```csharp
object a = true ? 1 : "1";//OK
var b = true ? 1 : "1";// CS0173
```
The above example indicates that the priority of object is extremely low. 
Even though it is currently available.

The same should apply to this [object].
Index is a value type, and all cases that require boxing should have lower priority.

**Reply by CyrusNajmabadi** [reply]
The API should accept a boxed Index.

**Reply by elgonzo** [reply]
@CyrusNajmabadi and the same API should also accept a boxed `Range`, then, right? How does this make any sense, that an `object`-based indexer is supposed to accept both `Index` as well as `Range`? Also, what exactly is the benefit of selecting an `object`-based indexer despite there being an accessible `int`-based indexer?

**Reply by CyrusNajmabadi** [reply]
> Also, what exactly is the benefit of selecting an `object`-based indexer despite there being an accessible `int`-based indexer?

It's consistent. It's what you get if you do `[1]` and `[^1]`, instead of having those call two separate indexers, despite being conceptually the same

**Reply by elgonzo** [reply]
Oh, so `[1]` calls the `int`-based indexer, and `[^1]` calls the `object`-based indexer, and the list pattern basically follows this established precedence. Gotcha. Thanks for the clarification. I am still not happy how it works, but at least i now understand the rationale behind it...

**Reply by CyrusNajmabadi** [reply]
If you want this to work for a particular api, I would recommend you all that API to add support. The language pieces are there. 

If the API is open source, you could even submit the change yourself. I'm guessing it would likely be accepted.

**Comment by dexcompiler** [comment]
Seems easy enough to manage this fluently using the `JsonDocument` API:

```c#
public static string? GetMessage(string json) =>
	JsonDocument.Parse(json).RootElement
		.GetProperty("entries")
		.EnumerateArray()
		.FirstOrDefault()
		.GetProperty("message")
		.GetString();
```
 Arugably more readable and easier on the eyes than the JS/Python/other languages above.

**Comment by CyrusNajmabadi** [comment]
Closing out as this has been answered. There are incredibly simple ways of working with Jason that people have already provided. And this is due to the support the language gives you. If you want particular apis to integrate better with the language support, you should take it up with the API authors. The language already has the tools to make the consumption as nice (or nicer) than the examples you provided from other languages.
